Basic float support

FurryAcetylCoA opened the pull request:
Tested with the following program
```c
#include <stdio.h>

int main(){
        int k=4;
        double y = k * 5.1;
        printf ("%f\n",y);
        return 0;
}
```
and
```c
#include <stdio.h>

int main(){
        float fa0 = .123f;
        float fa1 = 10.0f;
        float fa2 = fa0 * fa1;
        printf("%f\n", fa2);
}
```
`long double` is still not supported but others seems to be fine

FurryAcetylCoA committed:
Revert "Fix calling 64-bit types for riscv32"

This reverts commit cf9fbd9e6be2eaa1d09a81408ae2299f7525afe3.

FurryAcetylCoA committed:
Fix calling conv of int64 when exactly one register is available

FurryAcetylCoA committed:
fix split reg move in gfunc_call

Correct the following program
```
#include <stdio.h>

void f1(unsigned long long a){
    printf("%llx\n", a);
}

int main(){
    f1(0xc0ffeeff00001234);
    return 0;
}
```

FurryAcetylCoA committed:
wip: fix calling with double

now there are three way to represnent a register in tcc

FurryAcetylCoA committed:
complete basic float support

1. we should never set vtop->r/r2 to anything bigger than NB_REGS when
we want it to represnt an reg(whether it's aliased or not). Since gv()
will try to use vtop->r/r2 to index reg_classes[NB_REGS] and silently go
out of bounds
2. for gen_cvt_itof, the type of vtop after func call is the RETURN
type, not the type of argument

Tested with the following program
```
\#include <stdio.h>

int main(){
        int k=4;
        double y = k * 5.0;
        printf ("%f\n",y);
        return 0;
}
```

FurryAcetylCoA committed:
fix float to double convers

sellicott commented on riscv32-gen.c:
```diff
@@ -802,7 +809,7 @@ ST_FUNC void gfunc_call( int nb_args )
                     /* Only half of the last arg can be transferred by reg */
                     info[ i ] |= 16;
-                    stack_adj += 8;
+                    stack_adj += 4;
```
Could you change this to XLEN

sellicott commented on riscv32-gen.c:
```diff
@@ -941,16 +949,23 @@ ST_FUNC void gfunc_call( int nb_args )
+                // mv Ra+1, RR2
+                emit_MV( ireg(TREG_R(r2)) , ireg( vtop->r2 ) );
+                vtop->r2 = TREG_R(r2);
+            }
```
See comment below on the commit to tccgen.c

sellicott commented on tccgen.c:
```diff
@@ -319,11 +319,6 @@ static int RC2_TYPE(int t, int rc)
-#if defined TCC_TARGET_RISCV32 || defined TCC_TARGET_RISCV32
-    if (rc > RC_FLOAT)
-        return rc << 1;
-#endif
```
This code is why manual mv instructions are required to fix up two register argument passing. A better check is required so that if a non-existent register class is attempted, it gets pushed to the stack. That would eliminate the fix up code as well.

FurryAcetylCoA committed:
handle edge case of RC2_TYPE

When calling a function use the a0-a7 registers
Modify RC2_TYPE to upgrade the register class to the next one up
If no more register is available. Just use a temp reg and gfunc_call()
will store that on stack later
Remove fixup MV in gfunc_call()

Based on cf9fbd9

FurryAcetylCoA committed:
restore changes in gfunc_return

sellicott commented:
We will likely need to add a bit more in the tccgen code, but this looks good to me.
